pilot/model/config.go

```diff
@@ -497,7 +533,66 @@ func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destinat
 	return out
 }
 
+func (store *istioConfigStore) routeRulesV2(instances []*ServiceInstance, destination string) []Config {
+	out := make([]Config, 0)
+	configs, err := store.List(V1alpha2RouteRule.Type, NamespaceAll)
+	if err != nil {
+		return nil
+	}
+
+	for _, config := range configs {
+		rule := config.Spec.(*routingv2.RouteRule)
+
+		var found bool
+		for _, host := range rule.Hosts {
+			fqdn := ResolveFQDN(config.ConfigMeta, host)
+			if fqdn == destination {
+				found = true
+				break
+			}
+		}
+		if !found {
+			continue
+		}
+
+		if !matchSource(rule, instances) {
+			continue
+		}
+
+		out = append(out, config)
+	}
+
+	return out
+}
+
+// TODO: can the instance matching be optimized?
+func matchSource(rule *routingv2.RouteRule, instances []*ServiceInstance) bool {
```

Comment:
Yeah, we definitely need to do this on a per rule basis. Secondly, since we have settled on a sidecar per pod model, we can safely get rid of the array of instances and the associated for loop. That said, lets do the matchSource when applying individual rules in the array.